authz/grpc_audit_end2end_test.go

```diff
@@ -0,0 +1,304 @@
+				[
+					{
+						"name": "allow_all",
+						"request": {
+							"paths":
+							[
+								"*"
+							]
+						}
+					}
+				],
+				"deny_rules":
+				[
+					{
+						"name": "deny_all",
+						"request": {
+							"paths":
+							[
+								"/grpc.testing.TestService/StreamingInputCall"
+							]
+						}
+					}
+				],
+				"audit_logging_options": {
+					"audit_condition": "ON_DENY_AND_ALLOW",
+					"audit_loggers": [
+						{
+							"name": "stat_logger",
+							"config": {},
+							"is_optional": false
+						}
+					]
+				}
+			}`,
+			wantAllows: 2,
+			wantDenies: 1,
+		},
+		"Allow Unary - Audit Allow": {
+			authzPolicy: `{
+				"name": "authz",
+				"allow_rules":
+				[
+					{
+						"name": "allow_UnaryCall",
+						"request":
+						{
+							"paths":
+							[
+								"/grpc.testing.TestService/UnaryCall"
+							]
+						}
+					}
+				],
+				"audit_logging_options": {
+					"audit_condition": "ON_ALLOW",
+					"audit_loggers": [
+						{
+							"name": "stat_logger",
+							"config": {},
+							"is_optional": false
+						}
+					]
+				}
+			}`,
+			wantAllows: 2,
+			wantDenies: 0,
+		},
+		"Allow Typo - Audit Deny": {
+			authzPolicy: `{
+				"name": "authz",
+				"allow_rules":
+				[
+					{
+						"name": "allow_UnaryCall",
+						"request":
+						{
+							"paths":
+							[
+								"/grpc.testing.TestService/UnaryCall_Z"
+							]
+						}
+					}
+				],
+				"audit_logging_options": {
+					"audit_condition": "ON_DENY",
+					"audit_loggers": [
+						{
+							"name": "stat_logger",
+							"config": {},
+							"is_optional": false
+						}
+					]
+				}
+			}`,
+			wantAllows: 0,
+			wantDenies: 3,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			// Setup test statAuditLogger, gRPC test server with authzPolicy, unary and stream interceptors
+			lb := &loggerBuilder{
+				Stat: make(map[bool]int),
+			}
+			audit.RegisterLoggerBuilder(lb)
+			i, _ := NewStatic(test.authzPolicy)
+			cert, err := tls.LoadX509KeyPair(testdata.Path("x509/server1_cert.pem"), testdata.Path("x509/server1_key.pem"))
+			if err != nil {
+				t.Fatalf("tls.LoadX509KeyPair(x509/server1_cert.pem, x509/server1_key.pem) failed: %v", err)
+			}
+			ca, err := os.ReadFile(testdata.Path("x509/client_ca_cert.pem"))
+			if err != nil {
+				t.Fatalf("os.ReadFile(x509/client_ca_cert.pem) failed: %v", err)
+			}
+			certPool := x509.NewCertPool()
+			if !certPool.AppendCertsFromPEM(ca) {
+				t.Fatal("failed to append certificates")
+			}
+			creds := credentials.NewTLS(&tls.Config{
+				ClientAuth:   tls.RequireAndVerifyClientCert,
+				Certificates: []tls.Certificate{cert},
+				ClientCAs:    certPool,
+			})
+			s := grpc.NewServer(
+				grpc.Creds(creds),
+				grpc.ChainUnaryInterceptor(i.UnaryInterceptor),
+				grpc.ChainStreamInterceptor(i.StreamInterceptor))
+			defer s.Stop()
+			testgrpc.RegisterTestServiceServer(s, &testServer{})
+			lis, err := net.Listen("tcp", "localhost:0")
```

Comment:
Is it standard in these tests to run on `0`? Usually the standards are somewhere around `8xxx`

Reply:
By putting 0 here, the system will automatically bind a free port. This allows multiple tests running in parallel.